src/voter/mod.rs

```diff
@@ -420,6 +423,84 @@ fn instantiate_last_round<H, N, E: Environment<H, N>>(
 	}
 }
 
+/// Allows querying the state of the voter.
+pub trait VoterState<Id: Eq + std::hash::Hash> {
+	fn voter_state(&self) -> report::VoterState<Id>;
+}
+
+pub mod report {
+	use std::collections::{HashMap, HashSet};
+	use crate::weights::{VoteWeight, VoterWeight};
+
+	/// Basic data struct for the state of a round.
+	#[derive(PartialEq, Eq, Clone)]
+	#[cfg_attr(test, derive(Debug))]
+	pub struct RoundState<Id: Eq + std::hash::Hash> {
+		pub total_weight: VoterWeight,
+		pub threshold_weight: VoterWeight,
+
+		pub prevote_current_weight: VoteWeight,
+		pub prevote_ids: HashSet<Id>,
+
+		pub precommit_current_weight: VoteWeight,
+		pub precommit_ids: HashSet<Id>,
+	}
+
+	/// Basic data struct for the current state of the voter in a form suitable
+	/// for passing on to other systems.
+	#[derive(PartialEq, Eq)]
+	#[cfg_attr(test, derive(Debug))]
+	pub struct VoterState<Id: Eq + std::hash::Hash> {
+		pub background_rounds: HashMap<u64, RoundState<Id>>,
+		pub best_round: (u64, RoundState<Id>),
+	}
+}
+
+pub struct Inner<H, N, E> where
```

Comment:
Exposing a `struct Inner` from `voter.rs` seems a little ambiguous. Likewise, implementation on `Arc<RwLock<...>>` directly seems difficult from a backwards-compatibility perspective.

Reply:
It's used in `struct Voter` as well, but it's called `inner` although there are other fields. This would not have to be exposed as public with a wrapper struct, meaning it could gain a more descriptive name than `Inner`.

Reply:
The first approach Jon tried used a struct, the problem was that this struct had to be typed on `H, N, E`, and we'd have to propagate these types throughout the substrate codebase, it was especially hairy because types like `Environment` are currently not explicitly typed anywhere but would have to be due to the RPC initialization process. In substrate this trait is used as a trait object to avoid the type proliferation.

Reply:
OK, I assumed something like that. Still, the API could be improved substantially with a wrapping struct.

Reply:
Thanks for the feedback @rphmeier :)
I think we can actually remove the `pub` from `Inner` if I'm not mistaken? I'll add the missing docs

Reply:
IMO I think it's better to keep the struct anyway (although not exported) to be able to lock inner as a whole for correctness (rather than best and background distinctly).

Reply:
Updated the docs and removed the `pub` from `Inner`.

> OK, I assumed something like that. Still, the API could be improved substantially with a wrapping struct.

Sorry, do you mean that the API could have been improved if using a simple struct instead of the trait would have been practical? Or that the API could be improved by creating a struct wrapping `Arc<RwLock<...>>` just so that we don't have to `impl` directly on `Arc<RwLock<...>>`?

I started sketching on the latter but I got stuck on finding a good name ... :/
Probably `VoterState` would be appropriate for such a struct, but then the trait would need to be renamed ... `VoterStateQueryable`? Doesn't roll of the tongue very well ...

Reply:
`struct SharedVoterState<...>(Arc<RwLock<Inner<...>>>);` and then you implement `VoterState` for `SharedVoterState`.

Reply:
Added the `SharedVoterState` to wrap `Arc<RwLock<Inner>>`